Use a dedicated LDAP host and port for background jobs if configured

come-nc opened the pull request:
Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Use a dedicated LDAP host and port for background jobs if configured

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

artonge commented on apps/user_ldap/lib/Connection.php:
```diff
@@ -594,12 +594,18 @@ private function establishConnection() {
 			$isOverrideMainServer = ($this->configuration->ldapOverrideMainServer
 				|| $this->getFromCache('overrideMainServer'));
-			$isBackupHost = (trim($this->configuration->ldapBackupHost) !== "");
+			$isBackupHost = (trim($this->configuration->ldapBackupHost) !== "")
```
Would it make sense to rename to `hasBackupHost` to make it more obvious what this variable contains ?

artonge commented on apps/user_ldap/lib/Connection.php:
```diff
@@ -594,12 +594,18 @@ private function establishConnection() {
 				|| $this->getFromCache('overrideMainServer'));
-			$isBackupHost = (trim($this->configuration->ldapBackupHost) !== "");
+			$isBackupHost = (trim($this->configuration->ldapBackupHost) !== "")
+				&& (!\OC::$CLI || !$this->configuration->ldapBackgroundHost);
```
Hard to read, and adds conditions unrelated to the backup host.
I would split into another variable and make the config check more explicit:

```suggestion
				$shouldUseBackgroundHost = \OC::$CLI && (trim($this->configuration->ldapBackgroundHost) !== "");
```

artonge commented on apps/user_ldap/lib/Connection.php:
```diff
@@ -611,7 +617,7 @@ private function establishConnection() {
 			//if LDAP server is not reachable, try the Backup (Replica!) Server
 			if ($isBackupHost || $isOverrideMainServer) {
```
And make this condition more explicit

```suggestion
			if (($isBackupHost || $isOverrideMainServer) && !shouldUseBackgroundHost) {
```

come-nc committed:
Make code clearer for background host

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
I pushed a commit to make code clearer, but it makes it clear that there is a logic flaw:

When in CLI, background host is used and backuphost is not used even if backgroundhost do not answer. But, if backuphost is used by web UI because a fallback, then it will be used in CLI as well because of the cache.

So we need to decide to either:

- [ ] Fallback to backuphost even for CLI if backgroundhost do not anwser
- [ ] Never use backuphost in CLI event if backgroundhost is not answering, we use no fallback

(These is considering there is a backgroundhost configured, if not we always fallback if needed to keep the same behavior as before)

come-nc commented:
(There is also the question what to do when backup host is forced from admin settings, should that apply to cli if there is a background host?)

blizzz commented:
>  Fallback to backuphost even for CLI if backgroundhost do not anwser

I would say so (provided that backuphost differs from background host)

come-nc commented:
> > Fallback to backuphost even for CLI if backgroundhost do not anwser
> 
> I would say so (provided that backuphost differs from background host)

Ok, makes sense to me to always fallback to the backup if first host is not answering.
It also means forcing backup applies everywhere, which is more intuitive.

The problem with that is that we’ll need to split cache into 2 vars, to not go to backup for CLI when main is down but backgroundHost is still working.
I’ll implement that.

come-nc committed:
Fixed backup host logic

Now forcing backup host applies to both main and background.
And background will fallback to backup if not responding.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
/rebase

come-nc commented:
rebased

blizzz reviewed (approved):
looks good
